[DJM-585] Collect logs on EMR clusters

aboitreaud opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Enable logs agent for EMR logs collection if `DD_EMR_LOGS_ENABLED` is set to true, in DJM `EMR` install script flavor.

Impacting changes for team #fleet-automation:
- The type of `InitConfig` field of `IntegrationConfig` struct is changed from `[]any` to `any`. With the current `[]any` type, an empty init config is marshalled as `init_config: []` that is breaking the rest of the config that we may write in the yaml. We believe the yaml parser of the Agent doesn't support `[]` as array empty state. Type `any` is marshaled `init_config: null` when empty which fixes the parsing issue. Currently, initConfig is not used anywhere in the installer script.
- the Agent doesn't have permission to read the log files on EMR by default, so we add `dd-agent` user to the `yarn` group that has the required rights on the log folder. That is done in `common.setup.go`.
- the `executeCommantWithTimemout` function is moved from djm-specific folder to `common.setup.go` as we need it to add `dd-agent` to the `yarn` group in this file, and leaving it in `emr.go` would cause cycle imports.



### Motivation

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

aboitreaud committed:
log collection on workers

aboitreaud committed:
Collect stderr +stdout, add tags

aboitreaud committed:
add processor to logs config

aboitreaud committed:
make processors field an array

aboitreaud committed:
Add logs config on master + workers

aboitreaud committed:
Update emr.go

aboitreaud committed:
omit empty init config

aboitreaud committed:
add dd-agent to yarn group

aboitreaud committed:
Add log parser

aboitreaud committed:
Set log config even if sudo command fails

aboitreaud committed:
add ddagent to yarn group in setup[

aboitreaud committed:
Update emr.go

aboitreaud committed:
Add out + err to span

aboitreaud committed:
Add out + err to span

aboitreaud committed:
Update emr.go

aboitreaud committed:
add ddagent to yarn group in setup[

aboitreaud committed:
Move executeCommand to common

aboitreaud committed:
Move executeCommand to common

aboitreaud committed:
fix imports order

aboitreaud committed:
fix processor

raphaelgavache commented on pkg/fleet/installer/setup/common/config.go:
```diff
@@ -162,17 +162,27 @@ type DatadogConfigInstallerRegistry struct {
 // IntegrationConfig represents the configuration for an integration under conf.d/
 type IntegrationConfig struct {
-	InitConfig []any                   `yaml:"init_config"`
+	InitConfig []any                   `yaml:"init_config,omitempty"`
```
do the yarn and spark agent integrations still work with the omitempty?
I might being mistaken but I remember it not working without the field
You could check from fleet UI the status of the integrations on the host or run a datadog-agent status command on them to confirm

aboitreaud replied:
Good catch, thanks you are right.
If the logs are not enabled, because I removed the omitempty, it seems that the yarn integration is not working. The integration folder is not created if there's nothing to write in the file

aboitreaud replied:
The adopted solution is to keep the `omitempty` tag but change the type from `[]any` to `any`. I updated the PR description with a more detailed explanation of this choice

raphaelgavache commented on pkg/fleet/installer/setup/common/setup.go:
```diff
@@ -119,6 +121,15 @@ func (s *Setup) Run() (err error) {
 	}
+	isEMRorDataproc := s.flavor == "emr" || s.flavor == "dataproc"
+	if isEMRorDataproc {
+		// Add dd-agent to the yarn group to give the Agent permission to read the Spark log files.
```
to avoid leaking flavors here, could you instead propagate on the setup []string ddAgentAdditional groups, and add them here

aboitreaud committed:
DdAgentAdditionalGroups in Setup

raphaelgavache commented on pkg/fleet/installer/setup/common/setup.go:
```diff
@@ -119,6 +122,14 @@ func (s *Setup) Run() (err error) {
 		return fmt.Errorf("failed to write install info: %w", err)
 	}
+	for _, group := range s.DdAgentAdditionalGroups {
+		// Add dd-agent user to additional group for permission reason
```
you should be explicit with the case "in particular to enable reading log files not world readable", makes it easier to understand for next dev

aboitreaud committed:
better comment

aboitreaud committed:
Fix spark integration config

aboitreaud committed:
Integration config typed any

So that it's marshalled as "" empty string instead of "[]" that breaks the log config

aboitreaud committed:
Remove processor from yaml

aboitreaud committed:
Remove unused log Processor

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -52,6 +49,22 @@ var (
+		{
+			Type:    "file",
+			Path:    hadoopLogFolder + "*/*/stdout",
+			Source:  "worker_logs",
+			Service: "hadoop-yarn",
+			Tags:    "log_source:stdout",
```
```suggestion
			Source:  "hadoop-yarn",
			Service: "emr-logs",
```

- source should be the name of the integration: `hadoop-yarn` seems like a good value
- service is more complex since multiple service can run on the same cluster, I was thinking of a global name like `emr-logs`

The filename (stderr vs stdout) should directly be capture directly by the agent

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -52,6 +49,22 @@ var (
+		{
+			Type:    "file",
+			Path:    hadoopLogFolder + "*/*/stderr",
+			Source:  "worker_logs",
+			Service: "hadoop-yarn",
+			Tags:    "log_source:stderr",
```
```suggestion
			Source:  "hadoop-yarn",
			Service: "emr-logs",
```
same as above

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -74,7 +87,7 @@ type extraEmrInstanceInfo struct {
 // SetupEmr sets up the DJM environment on EMR
 func SetupEmr(s *common.Setup) error {
-
+	s.DdAgentAdditionalGroups = append(s.DdAgentAdditionalGroups, "yarn")
```
should this one be enabled if logs are enabled ?

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -52,6 +49,22 @@ var (
 			Value: "true",
 		},
 	}
+	emrLogs = []common.IntegrationConfigLogs{
```
would it make sense to only declare it when `DD_EMR_LOGS_ENABLED` is present ? Could think of creating a function `enableEmrLogs()` to contain all the logs logic

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -98,6 +111,13 @@ func SetupEmr(s *common.Setup) error {
 	if isMaster {
 		setupResourceManager(s, clusterName)
 	}
+	// Add logs config to both Resource Manager and Workers
+	sparkIntegration := s.Config.IntegrationConfigs["spark.d/conf.yaml"]
+	if os.Getenv("DD_EMR_LOGS_ENABLED") == "true" {
+		s.Config.DatadogYAML.LogsEnabled = true
+		sparkIntegration.Logs = emrLogs
+	}
+	s.Config.IntegrationConfigs["spark.d/conf.yaml"] = sparkIntegration
```
could this be written in this way instead ?

```suggestion
	if os.Getenv("DD_EMR_LOGS_ENABLED") == "true" {
		s.Config.DatadogYAML.LogsEnabled = true
		s.Config.IntegrationConfigs["spark.d/conf.yaml"] = emrLogs
	}
```

aboitreaud committed:
move all logs config in separate func

aboitreaud committed:
Update emr.go

paul-laffon-dd commented on pkg/fleet/installer/setup/djm/emr.go:
```diff
@@ -98,6 +94,11 @@ func SetupEmr(s *common.Setup) error {
 	}
+	// Add logs config to both Resource Manager and Workers
+	if os.Getenv("DD_EMR_LOGS_ENABLED") == "true" {
+		enableEmrLogs(s)
```
nitpick: can we add a log to indicate that the installer is enabling logs collection ? It would make it easier to troubleshoot

something like "Enabling EMR logs collection based on env variable DD_EMR_LOGS_ENABLED=true"

aboitreaud committed:
Add logs for the customers

aboitreaud commented:
/merge
